File: c/0073_SetMatrixZeroes.c
// Problem: Set Matrix Zeroes (https://leetcode.com/problems/set-matrix-zeroes/)
// Difficulty: Medium
// Date: 2021-12-04

/*
Given an m x n integer matrix matrix, if an element is 0, set its entire row and
column to 0's, and return the matrix.

You must do it in place.
*/

void setZeroes(int** matrix, int matrixSize, int* matrixColSize) {
    bool first_row_zero = false;
    bool first_col_zero = false;

    // find zeroes, store rows and columns to be set zero in the first row and column
    for (int i = 0; i < matrixSize; i++) {
        for (int j = 0; j < *matrixColSize; j++) {
            if (matrix[i][j] == 0) {
                if (i == 0) first_row_zero = true;
                if (j == 0) first_col_zero = true;
                matrix[i][0] = 0;
                matrix[0][j] = 0;
            }
        }
    }

    // set zeroes for rows
    for (int i = 1; i < matrixSize; i++) {
        if (matrix[i][0] == 0) {
            for (int j = 0; j < *matrixColSize; j++) {
                matrix[i][j] = 0;
            }
        }
    }

    // set zeroes for columns except first
    for (int j = 1; j < *matrixColSize; j++) {
        if (matrix[0][j] == 0) {
            for (int i = 0; i < matrixSize; i++) {
                matrix[i][j] = 0;
            }
        }
    }

    if (first_row_zero) {
        for (int j = 0; j < *matrixColSize; j++) {
            matrix[0][j] = 0;
        }
    }
    if (first_col_zero) {
        for (int i = 0; i < matrixSize; i++) {
            matrix[i][0] = 0;
        }
    }
}

// Wrong Answer: 
// Use of INT_MAX to mark whether rows/columns need to be zero
// Incorrect logic to determine whether to replace first row/column with zeroes
// void setZeroes(int** matrix, int matrixSize, int* matrixColSize) {
//     // find zeroes, store rows and columns to be set zero in the first row and column
//     for (int i = 0; i < matrixSize; i++) {
//         for (int j = 0; j < *matrixColSize; j++) {
//             if (matrix[i][j] == 0) {
//                 matrix[i][0] = INT_MAX;
//                 matrix[0][j] = INT_MAX;
//             }
//         }
//     }

//     // set zeroes for rows except first
//     for (int i = 1; i < matrixSize; i++) {
//         if (matrix[i][0] == INT_MAX) {
//             for (int j = 0; j < *matrixColSize; j++) {
//                 matrix[i][j] = 0;
//             }
//         }
//     }

//     // set zeroes for columns
//     bool top_left_zero = matrix[0][0] == INT_MAX;
//     for (int j = 0; j < *matrixColSize; j++) {
//         if (matrix[0][j] == INT_MAX) {
//             for (int i = 0; i < matrixSize; i++) {
//                 matrix[i][j] = 0;
//             }
//         }
//         if (top_left_zero) matrix[0][j] = 0;
//     }
// }

File: py/0073_SetMatrixZeroes.py
# 73. Set Matrix Zeroes

from utils import chunk
from typing import List


class Solution:
    def setZeroes(self, matrix: List[List[int]]) -> None:
        ROWS, COLS = len(matrix), len(matrix[0])

        # use separate flags for the first row/col because they cannot be used to identify themselves
        firstRowZero = False
        firstColZero = False

        # set the whole row/col to zero as long as there is one zero
        # mark whether a row/col contains a zero using the 0th-index
        for i in range(ROWS):
            for j in range(COLS):
                if matrix[i][j] == 0:
                    firstRowZero = i == 0 or firstRowZero
                    firstColZero = j == 0 or firstColZero
                    matrix[i][0] = 0
                    matrix[0][j] = 0

        # replace the first row/col last to prevent overwriting the zero flags
        # alternatively: loop through all m*n cells and set to 0 if matrix[i][0] == 0 or matrix[0][j] == 0
        for i in range(1, ROWS):
            if matrix[i][0] == 0:
                for j in range(1, COLS):
                    matrix[i][j] = 0

        for j in range(1, COLS):
            if matrix[0][j] == 0:
                for i in range(1, ROWS):
                    matrix[i][j] = 0

        if firstRowZero:
            for j in range(COLS):
                matrix[0][j] = 0

        if firstColZero:
            for i in range(ROWS):
                matrix[i][0] = 0


if __name__ == "__main__":
    testSize = 1
    puzzles = [
        [[1, 1, 1], [1, 0, 1], [1, 1, 1]],
        [[0, 1, 2, 0], [3, 4, 5, 2], [1, 3, 1, 5]]
    ]
    for puzzle in chunk(puzzles, testSize):
        Solution().setZeroes(*puzzle)
        for row in puzzle[0]:
            print(row)
        print()
